django/db/models/lookups.py

```diff
@@ -255,6 +255,19 @@ def get_db_prep_lookup(self, value, connection):
         )
 
 
+class CastRHSMixin:
+    """
+    A helper to cast explicitly the rhs placeholder to the lhs value, if needed.
+    """
+
+    def as_postgresql(self, compiler, connection):
+        sql, params = self.as_sql(compiler, connection)
+        cast_type = self.lhs.field.cast_db_type(connection)
+        if cast_type and cast_type.endswith("]"):
+            sql = sql.replace("%s", f"%s::{cast_type}")
+        return sql, params
+
+
```

Comment:
Ditto here, types should be adapted in `get_db_prep_value` by the backend so it emits the cast if necessary.

Reply:
This does not seem to be needed anymore so I dropped it completely.